File: atividade4.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <string>
using namespace std;


class AutomationDevice {
protected:
    int id;
    string name;

public:
    AutomationDevice(int idParam, string nameParam) {
        id = idParam;
        name = nameParam;
    }

    virtual void read() = 0;
    virtual void write(double value) = 0;
    virtual void display_status() const = 0;

    virtual ~AutomationDevice() {}
};


class Sensor : public AutomationDevice {
private:
    double value;

public:
    Sensor(int idParam, string nameParam) : AutomationDevice(idParam, nameParam) {
        value = 0.0;
    }

    void read() override {
        value += 1.0; 
    }

    void write(double val) override {
       
    }

    void display_status() const override {
        cout << "Sensor " << id << " (" << name << ") | Valor: " << value << endl;
    }
};


class Actuator : public AutomationDevice {
private:
    double state;

public:
    Actuator(int idParam, string nameParam) : AutomationDevice(idParam, nameParam) {
        state = 0.0;
    }

    void read() override {
       
    }

    void write(double val) override {
        state = val;
    }

    void display_status() const override {
        cout << "Actuator " << id << " (" << name << ") | Estado: " << state << endl;
    }
};

File: atividade4.py
from abc import ABC, abstractmethod
import random

# ---------------------------
# Classe base abstrata
# ---------------------------
class AutomationDevice(ABC):
    def __init__(self, id: int, name: str):
        self._id = id
        self._name = name

    @abstractmethod
    def read(self):
        pass

    @abstractmethod
    def write(self, value):
        pass

    @abstractmethod
    def status(self):
        pass

# ---------------------------
# Sensor
# ---------------------------
class Sensor(AutomationDevice):
    def __init__(self, id: int, name: str):
        super().__init__(id, name)
        self._value = 0.0

    def read(self):
        self._value = random.uniform(20.0, 30.0)  # valor aleatório
        return self._value

    def write(self, value):
        # Sensor não faz nada ao escrever
        pass

    def status(self):
        print(f"Sensor {self._id} ({self._name}) | Valor: {self._value:.2f}")

# ---------------------------
# Actuator
# ---------------------------
class Actuator(AutomationDevice):
    def __init__(self, id: int, name: str):
        super().__init__(id, name)
        self._state = 0.0

    def read(self):
